pwiz_tools/Skyline/EditUI/AssociateProteinsDlg.cs

```diff
@@ -178,36 +190,84 @@ public void UseFastaFile(string file)
-                    foreach (var peptide in peptidesForMatching)
+                    var fastaRecords = ParseFastaFileWithProgress(fastaFile).Select((tuple, index) =>
+                        Tuple.Create(index, tuple.Item1, tuple.Item2));
+                    ParallelEx.ForEach(fastaRecords, recordTuple =>
                     {
-                        // TODO(yuval): does digest matter?
-                        if (fasta.Sequence.IndexOf(peptide.Peptide.Target.Sequence, StringComparison.Ordinal) < 0)
+                        int index = recordTuple.Item1;
+                        int progressValue = (int) (recordTuple.Item2 * 100 / streamLength);
+                        var seq = recordTuple.Item3;
+                        var fasta = new FastaSequence(seq.Name, null, null, seq.Sequence);
+                        var substringFinder = new SubstringFinder(seq.Sequence, maxLength);
+                        var matches = new List<IdentityPath>();
+                        foreach (var peptide in peptidesForMatching)
+                        {
+                            // TODO(yuval): does digest matter?
+                            if (!substringFinder.ContainsSubstring(GetPeptideSequence(peptide)))
+                            {
+                                continue;
+                            }
+
+                            matches.Add(peptide);
+                        }
+
+                        lock (proteinAssociations)
                         {
-                            continue;
+                            broker.CancellationToken.ThrowIfCancellationRequested();
+                            if (progressValue > broker.ProgressValue && progressValue <= 100)
+                            {
+                                broker.ProgressValue = progressValue;
+                            }
+                            if (matches.Count > 0)
+                            {
+                                proteinAssociations.Add(Tuple.Create(index, fasta, matches));
+                            }
                         }
-                        matches.Add(peptide);
-                    }
-                    if(matches.Count > 0)
-                        proteinAssociations.Add(new KeyValuePair<FastaSequence, List<PeptideDocNode>>(fasta, matches));
+                    });
+                });
+                if (longWaitDlg.IsCanceled)
+                {
+                    return null;
+                }
+            }
+
+            // Protein associations may be out of order because of multi-threading, so put them back in order.
+            return proteinAssociations.OrderBy(tuple => tuple.Item1).Select(tuple =>
+                new KeyValuePair<FastaSequence, List<IdentityPath>>(tuple.Item2, tuple.Item3)).ToList();
+        }
+
+        /// <summary>
+        /// Returns tuples of FastaData along with the position in the Stream that the fasta record was found.

```

Comment:
A little confusing naming and commenting. The progress is not in this function, but it is important that this function uses "yield return" to parse progressively. Your parallel processing would completely break if someone changed this to return an array. Probably worth making that clearer.